src/compiler/rust/canister_methods/src/hash_file.rs

```diff
@@ -0,0 +1,131 @@
+            // This limit must be the same as on the node side or else the hashes will not match
+            let mut buffer = vec![0; limit];
+            let bytes_read = std::io::Read::read(&mut file, &mut buffer);
+
+            let previous_hash = get_partial_file_hash(&path);
+
+            match bytes_read {
+                Ok(bytes_read) => {
+                    if bytes_read != 0 {
+                        let new_hash = hash_chunk_with(&buffer, previous_hash.as_ref());
+                        set_partial_hash(&path, new_hash);
+                        let delay = core::time::Duration::new(0, 0);
+                        let closure = move || hash_file_by_parts(path, position + bytes_read as u64);
+                        ic_cdk_timers::set_timer(delay, closure);
+                    } else {
+                        // No more bytes to hash, set as final hash for this file
+                        match previous_hash {
+                            Some(hash) => {
+                                set_file_hash(&path, hash);
+                                ic_cdk::println!("Finish hashing");
+                                clear_partial_hash(&path);
+                                ic_cdk::println!("Partial hash cleared");
+                            }
+                            None => ic_cdk::println!("WARNING: No hash was found for {}", path),
+                        }
+                    }
+                }
+                Err(err) => panic!("Error reading file: {}", err),
+            }
+        }
+
+        pub fn get_partial_file_hash(path: &String) -> Option<Vec<u8>> {
+            PARTIAL_FILE_HASHES.with(|file_hashes| file_hashes.borrow().get(path).cloned())
+        }
+
+        fn set_partial_hash(path: &String, hash: Vec<u8>) {
+            PARTIAL_FILE_HASHES.with(|file_hashes| file_hashes.borrow_mut().insert(path.clone(), hash));
+        }
+
+        fn clear_partial_hash(path: &String) {
+            PARTIAL_FILE_HASHES.with(|file_hashes| file_hashes.borrow_mut().remove(path));
+        }
+
+        fn clear_file_hash(path: &String) {
+            let mut file_hashes = load_hashes().unwrap();
+            file_hashes.remove(path);
+            save_hashes(&file_hashes).unwrap();
+        }
+
+        fn set_file_hash(path: &String, hash: Vec<u8>) {
+            let mut file_hashes = load_hashes().unwrap();
+            file_hashes.insert(path.clone(), hash);
+            save_hashes(&file_hashes).unwrap();
+        }
+
+        fn hash_chunk_with(data: &[u8], previous_hash: Option<&Vec<u8>>) -> Vec<u8> {
+            let mut h = <sha2::Sha256 as sha2::Digest>::new();
+            sha2::Digest::update(&mut h, data);
+            if let Some(hash) = previous_hash {
+                sha2::Digest::update(&mut h, hash);
+            }
+            sha2::Digest::finalize(h).to_vec()
+        }
+
+        fn load_hashes() -> Result<HashMap<String, Vec<u8>>, std::io::Error> {
```

Comment:
I think reading from the file can be made much simpler by using simpler std read functions

Reply:
done